Python/Module6_Testing/Hypothesis.md

```diff
@@ -0,0 +1,1047 @@
+   3. Permutations of the list `[1, 2, 3, 4]`
+
+</div>
+
+
+<!-- #region -->
+<div class="alert alert-info">
+
+**Reading Comprehension: Improving our tests using Hypothesis**
+
+We will be writing improved tests for the basic functions – `count_vowels` and `merge_max_mappings` – by leveraging Hypothesis.
+This reading comprehension question will require more substantial work than usual.
+That being said, the experience that we will gain from this will be well worth the work.
+Keep in mind that solutions are included at the end of this page, and that these can provide guidance if we get stuck.
+
+Part 1: Testing correctness by construction
+
+Write a Hypothesis-driven test for the `count_vowels`; include this test in `tests/test_basic_functions`.
+This is a test function where we can explicitly construct a string in parts: its non-vowel characters, non-y vowels, and y-vowels.
+And thus, by constructing a string with a known number of vowel and non-vowel characters, we can know what the output of `count_vowels` *should* be for that input, and we can thus test for correctness in this way.
+We will want to read about the [st.text()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.text) strategy to construct the different parts of the string.
+The standard library's built-in `string` module provides a string of all printable characters (`string.printable`).
+
+We should ask ourselves: how general are input strings that we are constructing? Are there regular patterns in the strings that might prevent our test from identifying edge case bugs in `count_vowels`?
+
+
+Part 2: Property-based testing
+
+Write a Hypothesis-driven test for `merge_max_mappings` ; include this test in `tests/test_basic_functions`.
+Here, we can't simply contrive the inputs to `merge_max_mappings` in a general way and know what its output should be – we would have to re-implement the function to do that.
+Instead, we should *test the expected properties* of the merged dictionary.
+For example, one such property is that the merged dictionary should only contain maximum values.
+Another property would be that all of the keys among the input dictionaries should be present in the merged dictionary.
+Take some time to think of other such properties that we should test for.
+Ultimately we want to arrive at a comprehensive set of properties to test for such that we can be confident that our merged dictionary is correct.
+
+We will want to use [st.dictionaries()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.dictionaries) to describe the inputs to `merge_max_mappings`.
+Although dictionary keys can be any hashable object, suffice it to use both integers and text for the keys, and integers for the dictionary values in this test, for simplicity's sake.
+
+
+**We must remember to temporarily mutate our original functions to verify that these tests can actually catch bugs!**
```

Comment:
Might want a link to the idea of mutation testing here?

Reply:
I tell the reader about mutation testing and link them to the cosmic ray library earlier. Here I just want them to do some simple by-hand mutating